---
output: github_document
---


```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "inst/extdata/README",
  out.width = "100%"
)
```

# snpCYP

## Description

The purpose of snpCYP is to streamline and automate the process of
reading and analyzing single nucleotide polymorphisms (SNP) on CYP genes
and to easily allow for exploratory analysis by graphically outputting the 
potential disruptions on drug metabolism by correlating the SNPs to known
or predicted catalytic activity from PolyPhen-2 and DIDB databases. The 
packet uniquely improves current ability of researchers to automate their 
work flow by adding non existing R language support for correlating sample’s 
SNP’s/haplotypes to potential disruption on drug metabolism. This project
was developed using R 4.0.2 on MacOs 11.6.


## Installation

To download the package:
```{r eval=FALSE}
require("devtools")
devtools::install_github("a-albuquerque/snpCYP", build_vignettes = TRUE)
library("snpCYP")
```

To run the shinyApp:
Under construction

## Overview

```{r eval=FALSE}
ls("package:snpCYP")
data(package="snpCYP")
```

There are **4** datasets that are able to be accessed by the user:

**1. PossibleCpGIslands** is a fabricated (fake) dataset, that is a DNA strand
in the form of a string of nucleotides, that can be used as an example dataset
to calculate the amount of CpG islands are present in a DNA strand.

**2. MethylationObservation** is a dataset retrieved from a study performed
on *Ceratina calcarata* at the University of Georgia, where mapping of 
methylation and quantification of gene expression was completed (GSE111611). 
The DNA sequence before and after bisulfite conversion are found in this 
dataset.

**3. BeforeBariatricSurgery** is a dataset retrieved from a study performed
at INSERM on overweight adults who would be recieving bariatric surgery 
(GSE66306). This dataset includes the patients RNAseq data before the surgery.

**4. AfterBariatricSurgery** is a dataset retrieved from a study performed
at INSERM on overweight adults who had recieved bariatric surgery. 
(GSE66306).This dataset includes the patients RNAseq data after the surgery.


There are **5** functions that are able to be accessed by the user:

**1. differencesInGeneExpression** allows for users to be able to compare two
different RNAseq data from two different DNA strands (preferably with 
different methylation) and see the *n* genes with the largest difference in 
expression between the two RNAseq datasets in a graph. The user is able to 
specify *n*, which is the number of genes shown in the graphical output.

**2. findCpGIslands** allows for users to be able to determine how many CpG
islands are present in their DNA strand, as well as view where the CpG islands
are in the DNA strand as this function will output the nucleotides of the
given strand with the CpG islands highlighted.

**3. findDifferencesInDNASequence** allows for users to be able to determine
which nucleotides between two strands of DNA are different. This function will
output the two DNA strands with the nucleotides that differ between the strands
highlighted.

**4. findMethylatedCytosines** allows for user to compare a DNA sequence before
and after bisulfite conversion, and determine which cytosine nucleotides are
methylated or not. This function will output the original DNA strand (before
bisulfite conversion) with the cytosines that are methylated highlighted.

**5. runMethylExpress** allows for the user to run the shiny app for MethylExpress
so that they can interact with all of the functionalities of the package
through a GUI instead of the R commandline.

For more details about the functions, please take a look at the vignette for
this package:

```{r eval=FALSE}
browseVignettes("snpCYP")
```

The package tree structure is provided below:

``` r
- MethylExpress
  |- MethylExpress.Rproj
  |- DESCRIPTION
  |- NAMESPACE
  |- LICENSE
  |- README
  |- data
    |- AfterBariatricSurgery.rda
    |- BeforeBariatricSurgery.rda
    |- MethylationObservation.rda
    |- PossibleCpGIslands.rda
  |- inst
    |- extdata
      |- MethylExpressOverview.png
    |- shiny-scripts
      |- app.R
  |- man
    |- AfterBariatricSurgery.Rd
    |- BeforeBariatricSurgery.Rd
    |- countCpGIslands.Rd
    |- differencesInGeneExpression.Rd
    |- DNASequenceHighlights.Rd
    |- findCpGIslands.Rd
    |- findDifferenceOfExpression.Rd
    |- findDifferencesInDNASequence.Rd
    |- findMatchingSections.Rd
    |- findMethylatedCytosines.Rd
    |- highlight.Rd
    |- MethylationObservation.Rd
    |- observedAndExpected.Rd
    |- PossibleCpGIslands.Rd
    |- runMethylExpress.Rd
  |- R
    |- data.R
    |- differencesInGeneExpression.R
    |- findCpGIslands.R
    |- findDifferencesInDNASequence.R
    |- findMethylatedCytosines.R
    |- runMethylExpress.R
  |- vignettes
    |- differencesInGeneExpressionExResult.png
    |- findCpGIslandsExResult.png
    |- findDifferencesInDNASequenceExResult.png
    |- findMethylatedCytosinesExResult.png
    |- MethylExpressVignette.Rmd
    |- runMethylatedExpressExResult.png
  |- tests
    |- testthat.R
    |- testthat
    |- test-DifferencesInGeneExpression.R
    |- test-FindCpGIslands.R
    |- test-FindDifferencesInDNASequence.R
    |- test-FindMethylatedCytosines.R
```

An overview of the package is illustrated below 
(change this to refer to correct picture folder):
<!-- [](./inst/extdata/MethylExpressOverview.png) -->

## Contributions
The author of this package is Alcides Albuquerque.

For the function **DifferencesInGeneExpression**, I use the R package 
*ggplot2* to be able to plot the different amounts of gene expression.

For the functions **FindCpGIslands**, **FindDifferencesInDNASequence**, and
**FindMethylatedCytosines**, I use *rstudioapi* to be able to display my
markdown files (with the highlights) in the Rstudio Viewer panel.


## References
Adzhubei, I., Jordan, D. M., & Sunyaev, S. R. (2013). Predicting functional effect of human missense mutations using PolyPhen-2. Current Protocols in Human Genetics, 76 (1), 7–20.

Banerjee, P., Dunkel, M., Kemmler, E., & Preissner, R. (2020). SuperCYPsPred—a web server for the prediction of cytochrome activity. Nucleic Acids Research, 48 (W1), W580–W585. 

DIDB - the drug interaction database. (2021). In UW Drug Interaction Solutions. Washinton University. https://www.druginteractionsolutions.org/solutions/drug-interaction-database/

H. Wickham. ggplot2: Elegant Graphics for Data Analysis. Springer-Verlag New York, 2016.

Rostkowski, M., Spjuth, O., & Rydberg, P. (2013). WhichCyp: Prediction of cytochromes P450 inhibition. Bioinformatics, 29 (16), 2051–2052.

Wang, L.-L., Li, Y., & Zhou, S.-F. (2009). A bioinformatics approach for the phenotype prediction of nonsynonymous single nucleotide polymorphisms in human cytochromes P450. Drug Metabolism and Disposition, 37 (5), 977–991.

Zhang, T., Zhou, Q., Pang, Y., Wang, Y., Jin, C., Huo, J., Liu, L. A., & Wei, D. (2012). CYP-nsSNP: A specialized database focused on effect of non-synonymous SNPs on function of CYPs. Interdisciplinary Sciences: Computational Life Sciences, 4 (2), 83–89.


## Acknowledgements
This package was developed as part of an assessment for 2021 BCB410H: Applied Bioinformatics, University of Toronto, Toronto, CANADA

